peps/pep-0733.rst

```diff
@@ -0,0 +1,629 @@
+owning object is alive, but this time is hard to reason about. Such
+functions should not exist in the API without a mechanism that can
+make them safe.
+
+For containers, the API is currently missing bulk operations on the
+references of contained objects. This is particularly important for
+a stable ABI where ``INCREF`` and ``DECREF`` cannot be macros, making
+bulk operations expensive when implemented as a sequence of function
+calls
+[`Issue 15 <https://github.com/capi-workgroup/problems/issues/15>`__].
+
+Type Definition and Object Creation
+-----------------------------------
+
+The C API has functions that make it possible to create incomplete
+or inconsistent Python objects, such as ``PyTuple_New`` and
+``PyUnicode_New``. This causes problem when the object is tracked
+by GC or its ``tp_traverse``/``tp_clear`` functions are called.
+Such functions should be removed from the C API. Related functions,
+such as ``PyTuple_SetItem`` which is used to modify a partially
+initialized tuple, should also be removed (tuples are immutable
+once fully initialized)
+[`Issue 56 <https://github.com/capi-workgroup/problems/issues/56>`__].
+
+We identified a few issues with type definition APIs. For legacy
+reasons, there is often a significant amount of code duplication
+between ``tp_new`` and ``tp_vectorcall``
+[`Issue 24 <https://github.com/capi-workgroup/problems/issues/24>`__].
+The type slot function should be called indirectly, so that their
+signatures can change to include context information
+[`Issue 13 <https://github.com/capi-workgroup/problems/issues/13>`__].
+Several aspects of the type definition and creation process are not
+well defined, such as which stage of the process is responsible for
+initializing and clearing different fields of the type object
+[`Issue 49 <https://github.com/capi-workgroup/problems/issues/49>`__].
+
+Error Handling
+--------------
+
+Error handling in the C API is based on the error indicator which is stored
+on the thread state (in global scope). The design intention was that each
+API function returns a value indicating whether an error has occurred (by
+convention, ``-1`` or ``NULL``). When the program knows that an error
+occurred, it can fetch the exception object which is stored in the
+error indicator. We identified a number of problems which are related
+to error handling, pointing at APIs which are too easy to use incorrectly.
+
+There are functions that do not report all errors that occur while they
+execute. For example, ``PyDict_GetItem`` clears any errors that occur
+when it calls the key's hash function, or while performing a lookup
+in the dictionary
+[`Issue 51 <https://github.com/capi-workgroup/problems/issues/51>`__].
+
+Python code never executes with an in-flight exception (by definition),
+and by the same token C API functions should never be called with the error
+indicator set. This is currently not checked in most C API functions, and
```

Comment:
I think this is presuming a design that we haven't really discussed. While it's bad for an API to lose existing exception information, there's no reason to say "should never" here at this stage (and obviously I don't think we should say "never" at any stage, but we can discuss it).

How about:
```suggestion
and typically code using native functions should also be interrupted by
an error being raised. This is not checked in most C API functions, and
```